src/serializers/type_serializers/union.rs

```diff
@@ -70,22 +70,25 @@ impl UnionSerializer {
 
 impl_py_gc_traverse!(UnionSerializer { choices });
 
-fn to_python(
-    value: &Bound<'_, PyAny>,
-    include: Option<&Bound<'_, PyAny>>,
-    exclude: Option<&Bound<'_, PyAny>>,
+fn union_serialize<S, R>(
+    // if this returns `Ok(v)`, we picked a union variant to serialize, where
+    // `S` is intermediate state which can be passed on to the finalizer
+    mut selector: impl FnMut(&CombinedSerializer, &Extra) -> PyResult<S>,
+    // if called with `Some(v)`, we have intermediate state to finish
+    // if `None`, we need to just go to fallback
+    finalizer: impl FnOnce(Option<S>) -> R,
```

Comment:
I'm beginning to wonder if `finalizer` was a mistake; maybe we should remove it and `union_serialize` should instead return `PyResult<Option<S>>`. I think that would make the call paths more uniform as well as make it more consistent with `tagged_union_serialize`, which just takes `Option<S>`.

In `to_python` and `json_key` there would then be no special extra handling, whereas `serde_serialize` would then be able to do the follow-up `infer_serialize` step.

Reply:
Great call! I've implemented your suggestions!